### DISCUSSION TITLE
useEffect not updating with array

### DISCUSSION THREAD

**Discussion by davetapley** [mobxjs/mobx#2807] [Show and tell]
I've now convinced myself this isn't a bug, but I created a repro ⬇️  in a sandbox prior to realizing what was happening, so I thought others might benefit from seeing it 👀 

https://codesandbox.io/s/mobx-observable-array-with-react-lfh61?file=/src/App.tsx

---

The real gotcha for me was that I'd incorrectly _assumed_ that MobX would bring _some magic_ that would allow changes to the array to cause the `useEffect` to trigger.

I don't know why I assumed this 🤪

**Comment by iChenLei** [comment]
### simple debug
```js
// packages/react-reconciler/src/ReactFiberHooks.new.js
function updateEffectImpl(fiberFlags, hookFlags, create, deps): void {
  const hook = updateWorkInProgressHook();
  const nextDeps = deps === undefined ? null : deps;
  let destroy = undefined;

  if (currentHook !== null) {
    const prevEffect = currentHook.memoizedState;
    destroy = prevEffect.destroy;
    if (nextDeps !== null) {
      const prevDeps = prevEffect.deps;
      if (areHookInputsEqual(nextDeps, prevDeps)) {   👈🏻   // <- Here is the key of your problem
        hook.memoizedState = pushEffect(hookFlags, create, destroy, nextDeps);
        return;
      }
    }
  }
```

![2021-02-17 14 29 21](https://user-images.githubusercontent.com/14012511/108164863-93f8fc00-712c-11eb-8fc4-4718c6ba3521.png)

You can noticed that the `nextDeps` and `prevDeps` is same. So `React` think `deps` isn't changed so don't fire `effect`.

### Value vs. Reference
When you change the array value , the `nextDeps` and `prevDeps`'s value is still same cause they use same `Reference`.

### solution
```diff
+ const arr = array.slice(0);
  useEffect(() => {
    console.log(arr);
  }, [arr]);
```
or other `immutable` way .

> btw, this is not mobx's problem, it's something about `immutable` 、`reference` and javascript self. hope it's useful for you.

**Comment by mweststrate** [comment]
See also: https://mobx.js.org/react-integration.html#useeffect

**Reply by davetapley** [reply]
@mweststrate so the theory makes sense, but I'm getting weird results.

I've [forked my sandbox](https://codesandbox.io/s/mobx-observable-array-with-react-autorun-scgxv?file=/src/App.tsx) to use `autorun` and the empty dependency array `[]` as given in that link [here](https://github.com/mobxjs/mobx/blame/680ad86d76ab9dd9fbaeddcc73409047e08febca/docs/react-integration.md#L481-L486), but it:

1. Causes `eslint` error `React Hook useEffect has a missing dependency...` (which kinda makes sense), and:
2. Means neither the `bool` (i.e. by value) nor `array` (by reference) functions are triggered.

If I add the corresponding variable to the dependency array then we're back to the original behavior.

If I omit the dependency array argument to `useEffect` then it works, but [that feels wrong](https://stackoverflow.com/questions/58579426/in-useeffect-whats-the-difference-between-providing-no-dependency-array-and-an) 🤔

**Reply by iChenLei** [reply]
@davetapley Hi, sir. I think you need learn more concept about mobx. 

### Why your code not work.
1. `bool` case
```js
  useEffect(
    () =>
      autorun(() => {               // primitive values in JavaScript
        console.log(bool);  👈🏻    //  <-  and hence per definition not observable.
      }),
    []
  );
```
>All primitive values in JavaScript are immutable and hence per definition not observable. Usually that is fine, as MobX can just make the property that contains the value observable. In rare cases, it can be convenient to have an observable primitive that is not owned by an object. For such cases, it is possible to create an observable box that manages such a primitive.

2. `array` case
```js
  useEffect(
    () =>
      autorun(() => {
        console.log(array);
      }),
    []
  );
```
Yes, this is array and not primitive values. But still not work, what's wrong ?
What do `aoturun` do in the background ?

When we call a `autorun` func, we create a `new Reaction()` to `track` any observable deps, so we can know which `autorun` should be fire when the observable deps changed. 
```ts
export function autorun(
    view: (r: IReactionPublic) => any,
    opts: IAutorunOptions = EMPTY_OBJECT
): IReactionDisposer {
    /** ommited **/ 
    if (runSync) {
        // normal autorun
        reaction = new Reaction(
            name,
            function (this: Reaction) {
                this.track(reactionRunner)  👈🏻  // sync track
            },
            opts.onError,
            opts.requiresObservable
        )
    } else {
       /** ommited **/  👈🏻  // async track
    }
      /** ommited **/
    return reaction.getDisposer_()
}
```

In your case, we `console.log` the array, didn't do any `set` or `get` action, so the `aoturun` tracked zero deps.
The `Reaction`'s `observing_` is a empty array. Whatever you do for `array`, you still can't fire `autorun`(only first render, but that is `React`'s work).

![2021-02-18 02 01 01](https://user-images.githubusercontent.com/14012511/108247937-3f866880-718e-11eb-80d8-70e838e314e0.png)

Only change little code, it could work.  cause we do `get` action, so `Reaction` can track the `array`.

![021-02-18 02 17 26](https://user-images.githubusercontent.com/14012511/108248979-74df8600-718f-11eb-9fa3-e8543c23c072.png)

### Summary
No magic,  just use `ES6 Proxy` or `Object.defineproperty` to implement `mobx`'s reactive system. Javascript is a weird programming language, I am also confused sometimes. use `chrome debug` and read source code is useful for you to figure out why it work or why not work.

### Append
> 1. Causes eslint error React Hook useEffect has a missing dependency... (which kinda makes sense), and:

When you use mobx way to manage state, just add `// eslint-disable-next-line` to ignore it.

> If I omit the dependency array argument to useEffect then it works, but that feels wrong 

Yes, it's wrong. 
no deps args  =>  trigger by every re-render 
empty array deps => only trigger once.

**Reply by davetapley** [reply]
Thank you @iChenLei! That has explained a lot for me 🙏🏻 

I wonder why `mobx-state-tree` aren't using `box`, I will ask over there.

**Comment by iChenLei** [comment]
@davetapley sorry, sir. Maybe I am misleading you by my poor english. `mobx-state-tree` didn't do nothing wrong.
You get a `primitive values` cause you use [`Destructuring assignment` ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment), same with outdate mobx-react doc case (https://mobx-react.js.org/state-destruct)
```ts
const View = observer(({ model: { array, bool } }: ViewProps) => {  👈🏻
```
use case only mobx and mobx-state-tree https://jsfiddle.net/ichenlei/ace458dw/24/

previous commnet contain some mistake info, so I hide my comment.
